gk/fib.c

```diff
@@ -1426,7 +1492,7 @@ add_fib_entry_numerical(struct ip_prefix *prefix_info,
 	int ret;
 	struct gk_fib *neigh_fib;
 
-	if (prefix_info->len < 0)
+	if (prefix_info->len <= 0)
```

Comment:
A prefix length of zero is valid. If you try the command `route -n` on a Linux box, or analogous command on another platform, you'll like find an entry with `0.0.0.0/0`. Is it an issue of the LPM iterator, or an issue of DPDK's LPM?

Reply:
As commented in the previous review, `rte_lpm_rule_iterate()` (i.e. `rte_lpm_rule_iterate_v20()` and `rte_lpm_rule_iterate_v1604()`) checks variable `depth` with an `if` statement **and** with the macro `VERIFY_DEPTH`. Why do we need to test twice?

Reply:
I have tried to add a prefix with ``0.0.0.0/0`` in DPDK LPM, which failed. By checking the LPM source code, I find that ``if ((lpm == NULL) || (depth < 1) || (depth > RTE_LPM_MAX_DEPTH))``,  ``rte_lpm_add()`` fails when ``depth < 1``, also, ``rte_lpm_lookup()`` possibly returns ``-ENOENT on lookup miss``. So, I think DPDK LPM doesn't support that. Actually, with ``0.0.0.0/0``, which means a default next hop. It should be easily handled by the program who calls DPDK's LPM. Also, another issue may come up is to list all the rule entries, as we have to specify the ``ip address`` in current implementation, which will be used to do the match. One way to support ``listing all prefixes`` is to take the ip address parameter as a pointer, when the pointer is ``NULL`` we can return all the rules. What do you think?

Reply:
Let's stick with this DPDK's limitation. I've documented it in issue #91.

But given that this is an implementation limitation, `add_fib_entry_numerical()` and `del_fib_entry_numerical()` must log that Gatekeeper currently doesn't support default routes when it receives a prefix of length zero. This is important because default routes might be needed in production, and network administrators will need that message to figure out what's going on.

I think the LPM iterators should support zero-length prefixes as a way to list all prefixes in the LPM table. Being able to passing `NULL` in the IP address parameter of the IPv6 iterator would be nice, but notice that it's just a facility for the caller because a zero-length prefix means that no bit matter in the masked IP address, so the masked IP address already matches all prefixes.

Reply:
Replace the current documentation of parameter `depth` with the one below and remember to update the documentation of the IPv6 version as well:
```
* @param depth
*   Depth of the rule to be searched.
*   Pass zero to enumerate the whole LPM table.
```